enhanced storage resource textual summary page

OrGur1987 opened the pull request:
before:
![image](https://user-images.githubusercontent.com/106743023/216821519-d78db81f-f27f-4726-9d3b-20b384887480.png)

after:
![image](https://user-images.githubusercontent.com/106743023/216821564-fd00edf0-80bd-400b-bf2a-bfb5cc6dfca0.png)

OrGur1987 commented on app/controllers/storage_resource_controller.rb:
```diff
@@ -19,8 +19,14 @@ def show
+        {:title => _("Storage")},
+        # using url_controller in :url=> here doesn't work, '/storage_resource' doesn't route to '/storage_resource/show_list' for some reason
+        {:title => _("Storage Resources"), :url => '/storage_resource/show_list'},
+      ],
```
note the comment here..
using `{:title => _("Storage Resources"), :url => controller_url},` 
led to the error below, which seems to happen because '/storage_resource' doesn't route to '/storage_resource/show_list'.
couldn't find the bug preventing the routing, so hard-coded it instead.

the error:
```
[----] I, [2023-02-05T15:32:59.997413 #203714:15c70]  INFO -- : Started GET "/storage_resource" for 9.25.202.27 at 2023-02-05 15:32:59 +0200
[----] F, [2023-02-05T15:33:00.042269 #203714:15c70] FATAL -- :   
ActionController::RoutingError (No route matches [GET] "/storage_resource"):
  
actionpack (6.1.7.2) lib/action_dispatch/middleware/debug_exceptions.rb:33:in `call'
actionpack (6.1.7.2) lib/action_dispatch/middleware/show_exceptions.rb:33:in `call'
railties (6.1.7.2) lib/rails/rack/logger.rb:37:in `call_app'
railties (6.1.7.2) lib/rails/rack/logger.rb:28:in `call'
sprockets-rails (3.4.2) lib/sprockets/rails/quiet_assets.rb:13:in `call'
actionpack (6.1.7.2) lib/action_dispatch/middleware/remote_ip.rb:81:in `call'
actionpack (6.1.7.2) lib/action_dispatch/middleware/request_id.rb:26:in `call'
rack (2.2.6.2) lib/rack/method_override.rb:24:in `call'
rack (2.2.6.2) lib/rack/runtime.rb:22:in `call'
activesupport (6.1.7.2) lib/active_support/cache/strategy/local_cache_middleware.rb:29:in `call'
actionpack (6.1.7.2) lib/action_dispatch/middleware/executor.rb:14:in `call'
actionpack (6.1.7.2) lib/action_dispatch/middleware/static.rb:24:in `call'
rack (2.2.6.2) lib/rack/sendfile.rb:110:in `call'
actionpack (6.1.7.2) lib/action_dispatch/middleware/host_authorization.rb:148:in `call'
secure_headers (3.9.0) lib/secure_headers/middleware.rb:12:in `call'
railties (6.1.7.2) lib/rails/engine.rb:539:in `call'
railties (6.1.7.2) lib/rails/railtie.rb:207:in `public_send'
railties (6.1.7.2) lib/rails/railtie.rb:207:in `method_missing'
puma (4.3.12) lib/puma/configuration.rb:228:in `call'
puma (4.3.12) lib/puma/server.rb:727:in `handle_request'
puma (4.3.12) lib/puma/server.rb:476:in `process_client'
puma (4.3.12) lib/puma/server.rb:332:in `block in run'
puma (4.3.12) lib/puma/thread_pool.rb:134:in `block in spawn_thread'
```

jeffibm commented:
Hey @OrGur1987 , Code looks good (other than the usage of instance variable usage in helper files) , but I don't have records under` Storage / Storage Resources` so couldn't see the changes In the summary page.

@MelsHyrule do you see any records In your local env?

Could you also add a label to the PR.

OrGur1987 commented:
> Hey @OrGur1987 , Code looks good (other than the usage of instance variable usage in helper files) , but I don't have records under` Storage / Storage Resources` so couldn't see the changes In the summary page.
> 
> @MelsHyrule do you see any records In your local env?
> 
> Could you also add a label to the PR.

hi, thanks
the instance variable usage in helper files was copied from existing files I based my change on. 
what should we use instead?

also, do u have a clue regarding the routing problem I mentioned above?

jeffibm commented:
> > Hey @OrGur1987 , Code looks good (other than the usage of instance variable usage in helper files) , but I don't have records under` Storage / Storage Resources` so couldn't see the changes In the summary page.
> > @MelsHyrule do you see any records In your local env?
> > Could you also add a label to the PR.
> 
> hi, thanks the instance variable usage in helper files was copied from existing files I based my change on. what should we use instead?
> 
> also, do u have a clue regarding the routing problem I mentioned above?

Regarding the instance variable alerts, I think it's ok for now since its being used similarly in other places. but, you can refer to the link from the alert - [Rails/HelperInstanceVariable](https://docs.rubocop.org/rubocop-rails/2.14/cops_rails.html#railshelperinstancevariable) -

usualy, the` GET /storage_resource` searches for `index` action inside the `storage_resource_controller`. and the URL might not be defined in `routes.rb`. We had encountered similar issues before.

OrGur1987 commented:
> 

ok, I've added `index` to `storage_resource` get actions in `routes.rb` and removed `storage_resource` from `routes_without_index`, and it seems to work.

but I think I should move the addition of storage services from this PR to #8617 
<img width="665" alt="image" src="https://user-images.githubusercontent.com/106743023/216927024-6d18629d-4011-45d3-ab5b-ce861ac066bf.png">

jeffibm commented:
> ok, I've added `index` to `storage_resource` get actions in `routes.rb` and removed `storage_resource` from `routes_without_index`, and it seems to work.
> 
> but I think I should move the addition of storage services from this PR to #8617 <img alt="image" width="665" src="https://user-images.githubusercontent.com/106743023/216927024-6d18629d-4011-45d3-ab5b-ce861ac066bf.png">

ok, also please for any other url references. like navigation menu etc..

OrGur1987 commented:
> > @MelsHyrule do you see any records In your local env?
> 
> I do not sadly, @OrGur1987 do you think you could make a data-dump of your local environment so we have some data to test against?

Do u usually use pg_dump for that?

MelsHyrule commented:
@OrGur1987 Yes a `pg_dump` would work perfectly

OrGur1987 commented:
> @OrGur1987 Yes a `pg_dump` would work perfectly

sent it to u privately on Gitter

MelsHyrule commented:
Thanks for the data-dump! Really appreciate it 😄 

Started testing out your branch and found an issue

![Screen Shot 2023-02-08 at 4 12 39 PM](https://user-images.githubusercontent.com/29209973/217652938-dec58ddc-ece3-4661-8ab1-8646092fbb66.png)

Which i tracked down to [these lines](https://github.com/ManageIQ/manageiq-ui-classic/pull/8631/files#diff-5d79f85ea7f83c5d5e76301aa67d72c60ed594f1d1e0fd47a3c54b93ecf6a573R70-R72)

```
  def textual_physical_storage
    @record.ext_management_system.physical_storages.find(@record.physical_storage_id)
  end
```

`@record.ext_management_system` is nil and thus breaks the page. 

![Screen Shot 2023-02-08 at 4 12 28 PM](https://user-images.githubusercontent.com/29209973/217652955-466b273c-847b-409a-b867-e761de17dc49.png)

I found a similar issue with [these lines](https://github.com/ManageIQ/manageiq-ui-classic/pull/8631/files#diff-5d79f85ea7f83c5d5e76301aa67d72c60ed594f1d1e0fd47a3c54b93ecf6a573R42-R45)
```
    volumes = @record.cloud_volumes.map do |v|
      storage_service = @record.ext_management_system.storage_services.find(v.storage_service_id)
      [textual_link(v), number_to_human_size(v.size), v.health_state, storage_service.name]
    end
```

  again `@record.ext_management_system` is nil 

![Screen Shot 2023-02-08 at 4 24 14 PM](https://user-images.githubusercontent.com/29209973/217654257-fd9d3672-882b-4db9-ad04-27103eb6cead.png)

OrGur1987 commented:
> Thanks for the data-dump! Really appreciate it 😄
> 
> Started testing out your branch and found an issue
> 
> ![Screen Shot 2023-02-08 at 4 12 39 PM](https://user-images.githubusercontent.com/29209973/217652938-dec58ddc-ece3-4661-8ab1-8646092fbb66.png)
> 
> Which i tracked down to [these lines](https://github.com/ManageIQ/manageiq-ui-classic/pull/8631/files#diff-5d79f85ea7f83c5d5e76301aa67d72c60ed594f1d1e0fd47a3c54b93ecf6a573R70-R72)
> 
> ```
>   def textual_physical_storage
>     @record.ext_management_system.physical_storages.find(@record.physical_storage_id)
>   end
> ```
> 
> `@record.ext_management_system` is nil and thus breaks the page.
> 
> ![Screen Shot 2023-02-08 at 4 12 28 PM](https://user-images.githubusercontent.com/29209973/217652955-466b273c-847b-409a-b867-e761de17dc49.png)
> 
> I found a similar issue with [these lines](https://github.com/ManageIQ/manageiq-ui-classic/pull/8631/files#diff-5d79f85ea7f83c5d5e76301aa67d72c60ed594f1d1e0fd47a3c54b93ecf6a573R42-R45)
> 
> ```
>     volumes = @record.cloud_volumes.map do |v|
>       storage_service = @record.ext_management_system.storage_services.find(v.storage_service_id)
>       [textual_link(v), number_to_human_size(v.size), v.health_state, storage_service.name]
>     end
> ```
> 
> again `@record.ext_management_system` is nil
> 
> ![Screen Shot 2023-02-08 at 4 24 14 PM](https://user-images.githubusercontent.com/29209973/217654257-fd9d3672-882b-4db9-ad04-27103eb6cead.png)

Is this a probable scenario? Can a record have no ems?

MelsHyrule commented:
> Is this a probable scenario? Can a record have no ems?

Yes it can, so better safe than sorry and implement checks for it. In this case doing 
```
  def textual_physical_storage
    return nil unless @record.ext_management_system.present?
    @record.ext_management_system.physical_storages.find(@record.physical_storage_id)
  end
```
is enough

OrGur1987 commented:
> > Is this a probable scenario? Can a record have no ems?
> 
> Yes it can, so better safe than sorry and implement checks for it. In this case doing
> 
> ```
>   def textual_physical_storage
>     return nil unless @record.ext_management_system.present?
>     @record.ext_management_system.physical_storages.find(@record.physical_storage_id)
>   end
> ```
> 
> is enough

sorry, I actually already got a review from Adam on a future enhancement PR of `storage_resource_helper/textual_summary.rb`, and he suggested there to change from `@record.ems...` to directly using  `@record.storage_service` and `v.storage_service`
see #8617

agrare commented on app/helpers/storage_resource_helper/textual_summary.rb:
```diff
@@ -35,6 +39,14 @@ def textual_logical_total
+  end
+
+  def textual_physical_storage
+    @record.physical_storage
```
Stupid question (non-UI dev here) but should this be a textual_link to the physical_storage?  Or are you showing properties from the physical_storage record directly here?

OrGur1987 replied:
I guess it should be a textual_link, though it actually works as a link anyway

agrare replied:
That is the way it is in the HostInitiatorHelper:
```
  def textual_physical_storage
    textual_link(@record.physical_storage)
  end
```

So yes I think it would be best to stick with that

OrGur1987 replied:
done.
but what value does this method add? just out of curiosity..

agrare replied:
It handles showing a link to the argument passed in.

agrare commented:
@OrGur1987 the routes spec failure looks legitimate
